### ISSUE TITLE
classSchema feature request & missing classes on <li> elements

### ISSUE THREAD

**Issue by nahiyan8** [carloluis/ssuggestor#56]
## Do you want to request a _feature_ or report a _bug_?
*   **feature**
*   ~bug~

I want to request a feature.

### What is the current behavior?
1. Using `ssugestor` in a Bootstrap 4 environment still makes it output HTML with classes relevant to Bootstrap 3, specifically `glyphicon`.

   More generally, there is no way to adjust the `className`s for each of the HTML elements in the module's output.

   For example, there is no way to configure the module if we didn't want `form-control` to be on the `<input>` element, or if instead of `glyphicon` we wanted to use `fa fa-window-close` (Font Awesome).

2. The active `<li>` element does not have a class of its own, so it's impossible to target it using your own CSS. In fact, no `<li>` element has a class.

### What is the expected behavior?
1. It would be nice if we could pass in an object to set each of the `className`s, like:

```jsx
<Suggestor
    list={['suggestion-a', 'suggestion-b', 'suggestion-c', '...', 'suggestion-z']}
    arrow
    close
    classSchema={{
        root: 'input-group',
        arrowExpand: 'flashing',
        arrowRemove: 'fa fa-window-close',
        list: 'dropdown-menu'
        itemInactive: 'text-muted',
        itemActive: 'text-primary'
    }}
/>
```

2. Just an idea: auto-detect what CSS framework is in effect and apply the relevant `classSchema` accordingly?

   Or have pre-made `classSchema` for Bootstrap 3, Bootstrap 4, etc., so we can just import that from `suggestor`, and apply it ourselves.

3. The `<li>` elements should have a specific class instead of just inline CSS. Bootstrap has `dropdown-item` for this exact purpose. Also, the active `<li>` element uses an `active` class (in addition to the `dropdown-item` it's already using).

   Furthermore, I stongly suggest removing almost all inline CSS, and replace them CSS classes for styling. This would allow users to customize much more freely, and also helps separate “style” from “layout”.

#### If this is a feature request, what is motivation or use case for changing the behavior?
I was using Bootstrap 4 and this module was exactly what I needed. But I realised it used Bootstrap 3 classes to style itself so I thought that could be fixed.

### Additional information
Relevant versions:
* `ssuggestor 0.9.1`.
* `bootstrap 4.1.1`

**Comment by carloluis** [comment]
Thanks for pointing this out. 
Removing inline styles is a pending task for some time now.
I'll work first on extending styles customization through a `classSchema` property.
- Could you suggest which classes you consider would be nice to provide as default values for each elements?
If you have more ideas and want to contribute, please do so. I really appreciate it.

**Comment by nahiyan8** [comment]
What the hell? I thought I wrote a reply to this... Ok this is gonna be a pain to re-write but here goes:

I think what you have right now is pretty good. So I have a couple of ideas:

1. Add classes specific to suggestor's name to apply the inline CSS with. You could use [BEM](http://getbem.com/)

**Comment by nahiyan8** [comment]
Aww man, I thought I wrote a reply to this yesterday! Guess I forgot to press send now I've lost the message contents :/. Guess I have to re-write it from memory now, but here goes:

I think what you have right now is pretty good. So I have a couple of ideas:

1. Always apply some classes specific to each element to apply the inline CSS with. You could use [BEM](http://getbem.com/) but I don't like BEM because it's kinda bloaty and the other frameworks don't use it. So here are my suggestions for the classes for that:

```jsx
{
    root: 'suggestor',
    field: 'suggestor-field',
    buttonToggle: 'suggestor-toggle',
    buttonRemove: 'suggestor-remove',
    listHidden: 'suggestor-list hidden',
    listShown: 'suggestor-list',
    itemInactive: 'suggestor-item',
    itemActive: 'suggestor-item active',
    itemHighlight: 'suggestor-highlight'
}
```

2. Now `classSchema` can be used to apply classes for each specific theme like Bootstrap 3 or Bootstrap 4.

   For Bootstrap 3:  
   ```jsx
   const defaultClassSchema = {
       root: 'input-group',
       field: 'form-control',
       buttonToggle: 'glyphicon glyphicon-triangle-bottom', /* or 'caret' */
       buttonRemove: 'glyphicon glyphicon-remove',
       listHidden: 'dropdown-menu',
       listShown: 'dropdown-menu',
       itemInactive: 'dropdown-item',
       itemActive: 'dropdown-item active',
       itemHighlight: ''
   }
   ```

   In Bootstrap 4, we don't have glyphicons. This is a bit problematic, but thankfully we do have some alternatives:
   * Toggle: There's a `dropdown-toggle` class that makes a small arrow. (Actually Bootstrap 3 also includes `caret` for the same purpose like I noted above.)
   * Remove: There's a `close` class that gives the style of a clickable icon. But it needs `&times;` inside the `<a></a>` or `<button></button>` HTML to work.

     The easiest way to achieve this is to add a CSS selector in your .scss like this: `.suggestor-remove.close::after { content: '×'; }`. That way, it only selects it when Bootstrap 4 is being used.

     Alternatively, make a `suggestor-remove-icon` class by just copying the CSS from `close`, then add: `.suggestor-remove-icon::after { content: '×'; }`, and apply the class for both BS3 and BS4's `buttonRemove`. Boom. This removes the dependency on glyphicons in BS3, if you also use `caret` for `buttonToggle`.

   For Bootstrap 4:
   ```jsx
   const defaultClassSchema = {
       root: 'input-group',
       field: 'form-control',
       buttonToggle: 'dropdown-toggle dropdown-toggle-split',
       buttonRemove: 'close',
       listHidden: 'dropdown-menu',
       listShown: 'dropdown-menu',
       itemInactive: 'dropdown-item',
       itemActive: 'dropdown-item active',
       itemHighlight: 'text-high'
   }
   ```

3. Keep the list always rendered, whether or not it's being shown or not. But just hide it using: `.suggestor-list.hidden { display: none; }`. This allows users (and also you) to override that bit of CSS and make list opening and closing transition effects. If the DOM gets removed suddenly, then transition effects don't work.

**Comment by carloluis** [comment]
@nahiyan8 sorry for delaying the package styling schema update, but I lacked the time to work on this.

I just published [ssuggestor@1.0.0](https://www.npmjs.com/package/ssuggestor) with a new `theme` property to facilitate custom styling for the package. Included in the package there are also two __json__ files with _class schemas_ you can use depending of your Bootstrap version.

Please, check and test the package' latest version when you have time. Probably you need to use your custom `json` schema for Bootstrap 4. I also expect some some feedback from you.

In the near future, I have the intention to completely remove the `Bootstrap` class dependencies and provide a single custom theme for styling. Of course, the idea is to leave the door open to use custom themes and styles (in particular those using Bootstrap or any other css framework classes).

Thanks for your comments and request.

Cheers!
